core/src/main/java/kafka/log/remote/RemoteLogManager.java

```diff
@@ -698,14 +700,24 @@ public Optional<FileRecords.TimestampAndOffset> findOffsetByTimestamp(TopicParti
                     && rlsMetadata.endOffset() >= startingOffset
                     && isRemoteSegmentWithinLeaderEpochs(rlsMetadata, unifiedLog.logEndOffset(), epochWithOffsets)
                     && rlsMetadata.state().equals(RemoteLogSegmentState.COPY_SEGMENT_FINISHED)) {
-                    return lookupTimestamp(rlsMetadata, timestamp, startingOffset);
+                    // cache to avoid race conditions
+                    List<LogSegment> segmentsCopy = new ArrayList<>(unifiedLog.logSegments());
```

Comment:
Raised an edge case here where we may miss returning the right segment while searching for the target timestamp. This can happen when the search operation moves from remote segments to local segments search but the local segment is moved to remote storage and deleted from local segments successfully before it is searched successfully.

Thanks @kamalcph for raising the KAFKA-17637 as we discussed offline. This can be taken up in a follow-up PR.